wsd/COOLWSD.cpp

```diff
@@ -1001,6 +1007,109 @@ class PrisonPoll : public TerminatingPoll
 static std::unique_ptr<PrisonPoll> PrisonerPoll;
 
+#ifdef __linux__
+#if !MOBILEAPP
+class InotifySocket : public Socket
+{
+public:
+    InotifySocket():
+        Socket(inotify_init1(IN_NONBLOCK))
+        , m_stopOnConfigChange(true)
+    {
+        if (getFD() == -1)
+        {
+            LOG_WRN("Inotify - Failed to start a watcher for the configuration, disabling "
+                    "stop_on_config_change");
+            m_stopOnConfigChange = false;
+            return;
+        }
+
+        watch(COOLWSD_CONFIGDIR);
+    }
+
+    /// Check for file changes, stop the server if we find any
+    void handlePoll(SocketDisposition &disposition, std::chrono::steady_clock::time_point now, int events) override;
+
+    int getPollEvents(std::chrono::steady_clock::time_point /* now */,
+                      int64_t & /* timeoutMaxMicroS */) override
+    {
+        return POLLIN;
+    }
+
+    bool watch(std::string configFile);
+
+private:
+    bool m_stopOnConfigChange;
+    int m_watchedCount = 0;
+};
+
+bool InotifySocket::watch(const std::string configFile)
+{
+    LOG_TRC("Inotify - Attempting to watch " << configFile << ", in addition to current "
+                                             << m_watchedCount << " watched files");
+
+    if (getFD() == -1)
+    {
+        LOG_WRN("Inotify - Trying to watch config file " << configFile
+                                                         << " without an inotify file descriptor");
+        return false;
+    }
+
+    int watchedStatus;
+    watchedStatus = inotify_add_watch(getFD(), configFile.c_str(), IN_MODIFY);
+
+    if (watchedStatus == -1)
+        LOG_WRN("Inotify - Failed to watch config file " << configFile);
+    else
+        m_watchedCount++;
+
+    return watchedStatus != -1;
+}
+
+void InotifySocket::handlePoll(SocketDisposition & /* disposition */, std::chrono::steady_clock::time_point /* now */, int /* events */)
+{
+    LOG_TRC("InotifyPoll - woken up. Reload on config change: "
+            << m_stopOnConfigChange << ", Watching " << m_watchedCount << " files");
+    if (!m_stopOnConfigChange)
+        return;
+
+    char buf[4096];
+    const struct inotify_event* event;
+    ssize_t len;
+
+    LOG_TRC("InotifyPoll - Checking for config changes...");
+
+    while (true)
+    {
+        len = read(getFD(), buf, sizeof(buf));
+
+        if (len == -1 && errno != EAGAIN)
+        {
+            // Some read error, EAGAIN is when there is no data so let's not warn for it
+            LOG_WRN("InotifyPoll - Read error " << std::strerror(errno)
+                                                << " when trying to get events");
+        }
+        else if (len == -1)
+        {
+            LOG_TRC("InotifyPoll - Got to end of data when reading inotify");
+        }
+
+        if (len <= 0)
+            break;
+
+        for (char* ptr = buf; ptr < buf + len; ptr += sizeof(struct inotify_event) + event->len)
```

Comment:
inotify_event is one of those structs which uses a "struct hack" for a variable length trailing member.

struct inotify_event {
...
uint32_t len;      /* Size of name field */
char     name[];   /* Optional null-terminated name */
};

so sizeof(struct inotify_event) doesn't actually skip over a full struct, you would need to factor in whatever "len" claims to do that.

Either way, I don't think you can be sure you even have a complete entry even for the the first iteration to be sure "name" is null terminated if the record happens to be too long to fit in that buf buffer?

Reply:
I've read man inotify now and withdraw my comment :-)